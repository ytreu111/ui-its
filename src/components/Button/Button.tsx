import React, { FC, useCallback, useMemo } from "react";
import { LoadingIcon } from "ui-its-icons";
import { ThemeProvider } from "styled-components";

import {
  ButtonSizeEnum,
  ButtonStyleEnum,
  ButtonType,
  ButtonTypeEnum,
  IButtonProps,
} from "./model/Button.model";
import { Styled } from "./style/Button.styled";
import theme from "styles/theme/theme";

const {
  StyledButton,
  LoadingIconWrap,
  TextWrap,
  PrimaryButton,
  DangerButton,
  OrangeButton,
  YellowButton,
  GreenButton,
  PurpleButton,
  MagentaButton,
  BlueButton,
} = Styled

const resolveButton = (props: IButtonProps, type?: ButtonType) => {
  switch (type) {
    case ButtonTypeEnum.primary: {
      return <PrimaryButton {...props}/>
    }
    case ButtonTypeEnum.danger: {
      return <DangerButton {...props}/>
    }
    case ButtonTypeEnum.orange: {
      return <OrangeButton {...props}/>
    }
    case ButtonTypeEnum.yellow: {
      return <YellowButton {...props}/>
    }
    case ButtonTypeEnum.green: {
      return <GreenButton {...props}/>
    }
    case ButtonTypeEnum.purple: {
      return <PurpleButton {...props}/>
    }
    case ButtonTypeEnum.magenta: {
      return <MagentaButton {...props}/>
    }
    case ButtonTypeEnum.blue: {
      return <BlueButton {...props}/>
    }
    default:
      return <StyledButton {...props}/>
  }
}

const Button: FC<IButtonProps> = (
  {
    disabled,
    buttonStyle = ButtonStyleEnum.default,
    type,
    icon,
    size = ButtonSizeEnum.default,
    loading,
    onClick,
    clickEffect = true,
    children,
    rounded,
    ...props
  }) => {
  const onClickEffect = useCallback((event: MouseEvent) => {
    const button = event.currentTarget as HTMLElement;

    const circle = document.createElement("span");
    const diameter = Math.max(button?.clientWidth, button?.clientHeight);
    const radius = diameter / 2;
    const {x, y} = button.getBoundingClientRect();

    circle.style.width = circle.style.height = `${diameter}px`;
    circle.style.left = `${event.clientX - x - radius}px`;
    circle.style.top = `${event.clientY - y - radius}px`;
    circle.classList.add("wave");

    const wave = button.querySelector(".wave");

    if (wave) {
      wave.remove();
    }

    button.appendChild(circle);
  }, []);

  const onClickEvent = useCallback((e: MouseEvent) => {
    if (!loading) {
      if (clickEffect) onClickEffect(e);
      if (onClick) onClick(e);
    }
  }, [clickEffect, onClickEffect, onClick, loading]);

  const className = useMemo(() => {
    const className: Array<string> = [];

    if (disabled) className.push('disabled');

    if (buttonStyle && type) className.push(buttonStyle);

    if (size) className.push(size);

    if (rounded) className.push('rounded');

    if (loading) className.push('loading');

    return className.join(' ')
  }, [disabled, buttonStyle, type, size, rounded, loading]);

  const content = useMemo(() => {
    return (
      <>
        {icon}
        <TextWrap>{children}</TextWrap>
        {loading && <LoadingIconWrap><LoadingIcon/></LoadingIconWrap>}
      </>
    )
  }, [icon, children, loading]);

  const buttonProps = {
    ...props,
    className,
    children: content,
    onClick: onClickEvent,
  }

  return (
    <ThemeProvider theme={theme}>
      {resolveButton(buttonProps, type)}
    </ThemeProvider>
  )

}

export default Button;