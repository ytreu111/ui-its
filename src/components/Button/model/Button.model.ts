import { ReactNode } from "react";
import CSS from "csstype";

import IPadding from "types/interface/IPadding";
import IMargin from "types/interface/IMargin";

export interface IButtonProps extends IPadding, IMargin{
  type?: ButtonType;
  size?: SizeType;
  disabled?: boolean;
  buttonStyle?: StyleType;
  icon?: ReactNode;
  iconSize?: IIconSize | number;
  iconColor?: CSS.Property.Color;
  iconMargin?: number;
  loading?: boolean;
  onClick?: (e: MouseEvent) => void;
  clickEffect?: boolean;
  rounded?: boolean;
  children?: ReactNode;
  fontSize?: CSS.Property.FontSize;
  width?: CSS.Property.Width;
}

interface IIconSize {
  width: CSS.Property.Width;
  height: CSS.Property.Height;
}

export enum ButtonTypeEnum {
  primary = 'primary',
  danger = 'danger',
  orange = 'orange',
  yellow = 'yellow',
  green = 'green',
  purple = 'purple',
  magenta = 'magenta',
  blue = 'blue',
  transparent = 'transparent',
}

export enum ButtonSizeEnum {
  large = 'large',
  default = 'default',
  small = 'small',
}

export enum ButtonStyleEnum {
  //TODO Временно убрано по просьбе дизайнеров, тк оно не используется в системе
  // stroke = 'stroke',
  default = 'default',
  translucent = 'translucent',
  transparent = 'transparent',
}

export type ButtonType = keyof typeof ButtonTypeEnum;

export type SizeType = keyof typeof ButtonSizeEnum;

export type StyleType = keyof typeof ButtonStyleEnum;