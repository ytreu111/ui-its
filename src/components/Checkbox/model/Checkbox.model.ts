import { ChangeEvent, ReactElement } from "react";

export interface ICheckboxProps {
  checked?: boolean;
  value?: any;
  moved?: boolean;
  disabled?: boolean;
  checkboxSize?: number;
  name?: string;
  iconRender?: (checked: boolean) => ReactElement;
  onChange?: (checked: boolean, value: any, event: ChangeEvent<HTMLInputElement>) => void;
}