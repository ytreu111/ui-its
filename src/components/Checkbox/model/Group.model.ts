export interface ICheckboxOption {
  label: string;
  value: any;
  disabled?: boolean;
}

export interface ICheckboxGroupProps {
  options?: Array<ICheckboxOption>;
  value?: Array<any>,
  onChange?: (values: Array<any>) => void;
}