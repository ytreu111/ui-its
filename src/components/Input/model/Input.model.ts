import { ReactNode } from 'react';

export interface InputProps {
  value?: string;
  disabled?: boolean;
  small?: boolean;
  placeholder?: string;
  editableHelp?: string;
  label?: string;
  icon?: ReactNode;
  error?: boolean;
  autoWidth?: boolean;
  onChange?: (value: string) => void;
  onBlur?: () => void;
  onFocus?: () => void;
  name?: string;
  width?: number | string;
  optional?: () => void;
  optionalIcon?: ReactNode;
}
