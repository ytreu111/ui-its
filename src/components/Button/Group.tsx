import React, { FC, ReactElement, ReactNode, useCallback, useEffect, useMemo, useState } from "react";
import _ from "underscore";
import { StarIcon } from "ui-its-icons";

import { IButtonGroupProps, IButtonOptions } from "./model/Group.model";
import { ButtonStyleEnum, ButtonTypeEnum } from "./model/Button.model";
import Button from "./Button";
import { Styled } from "./style/Group.styled";

const {
  GroupWrap,
} = Styled;

const getButtonComponent = (button: ReactElement, buttonOptions: IButtonOptions, currentValue?: string) => {
  const props = button.props;
  const selectButton = props.value && props.value === currentValue;
  const currentButtonType = props.type ?? ButtonTypeEnum.primary;
  const buttonIcon = props.icon ? props.icon : selectButton ? undefined : buttonOptions.defaultIcon;
  const buttonKey = button.key ?? props.value ?? _.uniqueId();
  const className = button.props?.className?.split(' ') ?? [];

  let buttonChildren;
  let currentButtonStyle;

  className.push('group__button')

  if (selectButton) {
    currentButtonStyle = buttonOptions.selectButton;
    buttonChildren = props.children;
    className.push('group__button--active');
  } else {
    currentButtonStyle = buttonOptions.deselectButton;
    buttonChildren = null;
  }

  const buttonProps = {
    ...button.props,
    buttonStyle: currentButtonStyle,
    type: currentButtonType,
    children: buttonChildren,
    icon: buttonIcon,
    onClick: buttonOptions.onChange,
    className: className.join(' '),
  }

  return <Button key={buttonKey} {...buttonProps}/>
}

const childrenButton = (children: ReactNode, buttonOptions: IButtonOptions, currentValue?: string, content?: Array<ReactNode>) => {
  let newContent = content ? [...content] : [];

  if (children && typeof children === 'object') {
    if (Array.isArray(children)) {
      for (const element of children) {
        if (element.type === React.Fragment) {
          newContent = childrenButton(element.props.children, buttonOptions, currentValue, newContent);
        } else if (element.type === Button) {
          const button = getButtonComponent(element, buttonOptions, currentValue);
          newContent.push(button);
        }
      }
    } else {
      if ('type' in children) {
        if (children.type === React.Fragment) {
          newContent = childrenButton(children.props.children, buttonOptions, currentValue, newContent);
        } else if (children.type === Button) {
          const button = getButtonComponent(children, buttonOptions, currentValue);
          newContent.push(button);
        }
      }
    }
  }

  return newContent;
}

const Group: FC<IButtonGroupProps> = (
  {
    value,
    buttonGap = 8,
    selectButton = ButtonStyleEnum.default,
    deselectButton = ButtonStyleEnum.translucent,
    defaultIcon = <StarIcon/>,
    onChange,
    defaultValue,
    children,
  }) => {
  const [groupValue, setGroupValue] = useState<string>();

  useEffect(() => {
    if (defaultValue) setGroupValue(defaultValue);
  }, [defaultValue]);

  const onChangeButton = useCallback((event, value) => {
    if (onChange) onChange(event, value);
    setGroupValue(value);
  }, [onChange]);

  const content = useMemo(() => {
    if (children) {
      const currentValue = value ?? groupValue;
      const buttonOptions = {
        selectButton,
        deselectButton,
        defaultIcon,
        onChange: onChangeButton,
      }

      return childrenButton(children, buttonOptions, currentValue);
    }

    return null;
  }, [children, defaultIcon, deselectButton, groupValue, onChangeButton, selectButton, value])

  return (
    <GroupWrap gap={buttonGap}>
      {content}
    </GroupWrap>
  );
}

export default Group;