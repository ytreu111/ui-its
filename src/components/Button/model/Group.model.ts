import { StyleType } from './Button.model';
import { ReactNode } from 'react';

export interface IButtonGroupProps {
  value?: string;
  buttonGap?: number;
  selectButton?: StyleType,
  deselectButton?: StyleType,
  defaultIcon?: ReactNode,
  onChange?: (event: MouseEvent, value?: string) => void;
  defaultValue?: string;
}

type ButtonOptionsFields = 'selectButton' | 'deselectButton' | 'defaultIcon' | 'onChange';

export type ButtonOptions = Required<Pick<IButtonGroupProps, ButtonOptionsFields>>;