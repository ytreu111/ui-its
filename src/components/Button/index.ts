import { FC } from "react";

import InternalButton from "./Button";
import Group from './Group'
import { IButtonProps } from "./model/Button.model";

export interface CompoundedComponent extends FC<IButtonProps> {
  Group: typeof Group;
}

const Button = InternalButton as CompoundedComponent;

Button.Group = Group;

export type { IButtonProps, ButtonStyleEnum, ButtonSizeEnum, ButtonTypeEnum } from './model/Button.model'
export type { IButtonGroupProps } from './model/Group.model'
export default Button;