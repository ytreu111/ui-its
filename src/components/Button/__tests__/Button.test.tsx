import { render, screen, prettyDOM } from '@testing-library/react';
import Button from '../Button';
import { ButtonType, ButtonTypeEnum } from '../model/Button.model';
import { ButtonTestElemsEnum } from './Button.test.model';
import userEvent from '@testing-library/user-event';

const onClick = jest.fn();

describe('Компонент Button', () => {
  test('Указание типа', () => {
    const { rerender } = render(<Button />);

    expect(screen.getByTestId(ButtonTypeEnum.primary)).toBeInTheDocument();

    for (let buttonType of Object.values(ButtonTypeEnum)) {
      rerender(<Button type={buttonType as ButtonType} />);
      expect(screen.getByTestId(buttonType)).toBeInTheDocument();
    }
  });
  test('Наличие иконки загрузки', () => {
    render(<Button loading />);
    expect(screen.getByTestId(ButtonTestElemsEnum.loading)).toBeInTheDocument();
  });
  test('Эффект при нажатии', () => {
    render(<Button clickEffect onClick={onClick}/>);
    const button = screen.getByRole('button');
    userEvent.click(button);
    userEvent.click(button);
    const wave = screen.getAllByTestId(ButtonTestElemsEnum.wave);
    expect(wave.length).toBe(1);
    expect(onClick).toBeCalledTimes(2);
  });
  test('ClassName', () => {
    const { baseElement, rerender } = render(<Button className={'test'}/>);
    expect(prettyDOM(baseElement)).toMatchSnapshot();
    rerender(<Button disabled rounded>button</Button>)
    expect(prettyDOM(baseElement)).toMatchSnapshot();
  });
});