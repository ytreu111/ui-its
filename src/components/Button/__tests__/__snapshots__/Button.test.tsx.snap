// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`–ö–æ–º–ø–æ–Ω–µ–Ω—Ç Button ClassName 1`] = `
"[36m<body>[39m
  [36m<div>[39m
    [36m<button[39m
      [33mclass[39m=[32m\\"sc-dkPtRN sc-hKwDye ifkkxM enJpyK test size--default empty\\"[39m
      [33mdata-testid[39m=[32m\\"primary\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-gsDKAQ\\"[39m
      [36m/>[39m
    [36m</button>[39m
  [36m</div>[39m
[36m</body>[39m"
`;

exports[`–ö–æ–º–ø–æ–Ω–µ–Ω—Ç Button ClassName 2`] = `
"[36m<body>[39m
  [36m<div>[39m
    [36m<button[39m
      [33mclass[39m=[32m\\"sc-dkPtRN sc-hKwDye ifkkxM enJpyK disabled size--default rounded\\"[39m
      [33mdata-testid[39m=[32m\\"primary\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-gsDKAQ\\"[39m
      [36m>[39m
        [0mbutton[0m
      [36m</div>[39m
    [36m</button>[39m
  [36m</div>[39m
[36m</body>[39m"
`;
